server/src/main/java/org/elasticsearch/search/internal/InternalSearchResponse.java

```diff
@@ -43,32 +44,66 @@ public static InternalSearchResponse empty(boolean withTotalHits) {
                                   SearchProfileShardResults profileResults, boolean timedOut, Boolean terminatedEarly,
                                   int numReducePhases) {
-        super(hits, aggregations, suggest, timedOut, terminatedEarly, profileResults, numReducePhases);
+        super(hits, suggest, timedOut, terminatedEarly, profileResults, numReducePhases);
+        this.hasAggregations = aggregations != null;
+        this.aggregations = aggregations;
     }
 
-    public InternalSearchResponse(StreamInput in) throws IOException {
-        super(
-                new SearchHits(in),
-                in.readBoolean() ? new InternalAggregations(in) : null,
-                in.readBoolean() ? new Suggest(in) : null,
-                in.readBoolean(),
-                in.readOptionalBoolean(),
-                in.readOptionalWriteable(SearchProfileShardResults::new),
-                in.readVInt()
-        );
+    public static InternalSearchResponse read(StreamInput in) throws IOException {
+        SearchHits hits = new SearchHits(in);
+        /*
+         * We don't delay reading the aggs here because it'd do two things:
+         * 1. Delay reading aggs for cross cluster search until all clusters
+         *    are ready to be reduced. This is almost certainly a good thing.
+         * 2. Delay reading aggs in the any *other* time we serialize the
+         *    search response.
+         * We haven't done the digging to be sure that the second one is ok yet. 
+         */
+        DelayableWriteable<InternalAggregations> aggs = in.readOptionalWriteable(
+                si -> DelayableWriteable.referencing(new InternalAggregations(si)));
+        Suggest suggest = in.readOptionalWriteable(Suggest::new);
+        boolean timedOut = in.readBoolean();
+        Boolean terminatedEarly = in.readOptionalBoolean();
+        SearchProfileShardResults profileResults = in.readOptionalWriteable(SearchProfileShardResults::new);
+        int numReducePhases = in.readVInt();
+        return new InternalSearchResponse(hits, aggs, suggest, profileResults, timedOut, terminatedEarly, numReducePhases);
     }
 
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         hits.writeTo(out);
-        out.writeOptionalWriteable((InternalAggregations)aggregations);
+        out.writeOptionalWriteable(aggregations == null ? null : aggregations.get());
         out.writeOptionalWriteable(suggest);
         out.writeBoolean(timedOut);
         out.writeOptionalBoolean(terminatedEarly);
         out.writeOptionalWriteable(profileResults);
         out.writeVInt(numReducePhases);
     }
+
+    @Override
+    public InternalAggregations aggregations() {
+        if (hasAggregations) {
+            if (aggregations == null) {
+                throw new IllegalStateException("aggregations already consumed");
+            }
+            return aggregations.get();
```

Comment:
I am worried about the hidden cost of calling this on `serialized` aggs. Can we at least ensure that we don't get aggregations twice like [here](https://github.com/Kobar-Demo/Elasticsearch_Clone/blob/master/x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/MutableSearchResponse.java#L132) ?